Test-FileCatalog changes to use File.OpenRead to better handle when the file is executing or open Fix #20703

dxk3355 opened the pull request:
Test-FileCatalog changes to use File.OpenRead Fix #20703 
Added test cases to test Test-FileCatalog with an exe that is executing 
Added test cases to test Test-FileCatalog with files with FileShare.Read and ReadWrite

<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

<!-- Summarize your PR between here and the checklist. -->

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

Commit:
Test-FileCatalog improvements for when a file is open

Test-FileCatalog changes to use File.OpenRead Fix #20703
Added test cases to test Test-FileCatalog with an exe that is executing
Added test cases to test Test-FileCatalog with files with FileShare.Read and ReadWrite

iSazonov commented on src/System.Management.Automation/security/CatalogHelper.cs:
```diff
@@ -423,7 +423,7 @@ internal static string CalculateFileHash(string filePath, string hashAlgorithm)
             try
             {
-                fileStream = File.Open(filePath, FileMode.Open, FileAccess.Read);
+                fileStream = File.OpenRead(filePath);
```
Taking into account a nature of the issue we fix I'd prefer explicit parameters.
```suggestion
                fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
```

dxk3355 replied:
I was trying to be consistent with the existing code base.  Get-FileHash uses this method.  If this is the method to resolve it, I would then create an Issue that these are inconsistent.

iSazonov replied:
I believe the source of this bug is the fact that no explicit arguments can be seen. Therefore, I ask to specify the arguments explicitly.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1:
```diff
@@ -416,6 +416,62 @@ Describe "Test suite for NewFileCatalogAndTestFileCatalogCmdlets" -Tags "CI" {
+        It "Test-FileCatalog should pass when target file has an open reader with FileMode Open FileAccess Read and FileShare Read" {
+            $script:catalogPath = "$env:TEMP\TestCatalogFileOpenValidation.cat"
+            $null = New-FileCatalog -Path "$env:TEMP\testCatalog\" -CatalogFilePath $script:catalogPath -CatalogVersion 2
+            $fileHandle = [System.IO.File]::Open("$env:TEMP\testCatalog\test.txt", [System.IO.FileMode]::Open, [System.IO.FileAccess]::Read, [System.IO.FileShare]::Read)
```
$fileStream?

dxk3355 replied:
I'm trying to create an open file to verify that the Test-Catalog can hash an open file as this was an aspect of the issue and change.  Is there a reason to not have an open stream?

iSazonov replied:
I mean File.Open returns FileStream, not handle.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1:
```diff
@@ -416,6 +416,62 @@ Describe "Test suite for NewFileCatalogAndTestFileCatalogCmdlets" -Tags "CI" {
+                $result | Should -Be "Valid"
+            }
+            finally {
+                $fileHandle.Close()
```
If it is FileStream please remove.

Commit:
Changes requested from review

Change the OpenRead to Open with parameter for same effect
Rename fileHandle to fileStream

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1:
```diff
@@ -416,6 +416,62 @@ Describe "Test suite for NewFileCatalogAndTestFileCatalogCmdlets" -Tags "CI" {
+            Remove-Item "$env:TEMP\testCatalog" -Recurse -Force -ErrorAction SilentlyContinue
+        }
+
+        It "Test-FileCatalog should pass when target file has an open reader with FileMode Open FileAccess Read and FileShare Read" {
```
The two tests have identical code. Please use `-Case` parameter for FileShare Read/ReadWrite and remove second test.

dxk3355 replied:
The tests for Context "TestCatalog Skip Validation Tests" are basically having the same problem you're concerned about here as they only change a single parameter.

Commit:
Update FileCatalog.Tests.ps1

Move the FileShare to a parameterized test case per review

iSazonov commented:
@dxk3355 Thanks for your contribution!
